controller/konnect/ops/ops_adopt.go

```diff
@@ -0,0 +1,669 @@
+package ops
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"slices"
+	"sort"
+
+	sdkkonnectcomp "github.com/Kong/sdk-konnect-go/models/components"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	commonv1alpha1 "github.com/kong/kong-operator/api/common/v1alpha1"
+	kcfgkonnect "github.com/kong/kong-operator/api/konnect"
+	konnectv1alpha1 "github.com/kong/kong-operator/api/konnect/v1alpha1"
+	"github.com/kong/kong-operator/controller/konnect/constraints"
+	sdkops "github.com/kong/kong-operator/controller/konnect/ops/sdk"
+)
+
+// AdoptMatch performs adoption of a Konnect entity in "match" mode. When successful, the existing
+// entity in Konnect is managed by the operator without modifying it.
+func AdoptMatch[
+	T constraints.SupportedKonnectEntityType,
+	TEnt constraints.EntityType[T],
+](
+	ctx context.Context,
+	sdk sdkops.SDKWrapper,
+	cl client.Client,
+	ent TEnt,
+) error {
+	adoptable, ok := any(ent).(constraints.KonnectEntityTypeSupportingAdoption)
+	if !ok {
+		err := fmt.Errorf("%T does not support adoption", ent)
+		SetKonnectEntityAdoptedConditionFalse(ent, kcfgkonnect.KonnectEntitiesFailedToAdoptReason, err)
+		SetKonnectEntityProgrammedConditionFalse(ent, kcfgkonnect.KonnectEntitiesFailedToAdoptReason, err)
+		return err
+	}
+
+	adoptOpts := adoptable.GetAdoptOptions()
+	if adoptOpts == nil {
+		err := fmt.Errorf("adopt options must be provided")
+		SetKonnectEntityAdoptedConditionFalse(ent, kcfgkonnect.KonnectEntitiesFailedToAdoptReason, err)
+		SetKonnectEntityProgrammedConditionFalse(ent, kcfgkonnect.KonnectEntitiesFailedToAdoptReason, err)
+		return err
+	}
+
+	if adoptOpts.From != commonv1alpha1.AdoptSourceKonnect {
+		err := fmt.Errorf("unsupported adopt source %q", adoptOpts.From)
+		SetKonnectEntityAdoptedConditionFalse(ent, kcfgkonnect.KonnectEntitiesFailedToAdoptReason, err)
+		SetKonnectEntityProgrammedConditionFalse(ent, kcfgkonnect.KonnectEntitiesFailedToAdoptReason, err)
+		return err
+	}
+
+	if adoptOpts.Konnect == nil || adoptOpts.Konnect.ID == "" {
+		err := fmt.Errorf("konnect ID must be provided for adoption")
+		SetKonnectEntityAdoptedConditionFalse(ent, kcfgkonnect.KonnectEntitiesFailedToAdoptReason, err)
+		SetKonnectEntityProgrammedConditionFalse(ent, kcfgkonnect.KonnectEntitiesFailedToAdoptReason, err)
+		return err
+	}
+
+	mode := adoptOpts.Mode
+	if mode == "" {
+		mode = commonv1alpha1.AdoptModeMatch
+	}
+	if mode != commonv1alpha1.AdoptModeMatch {
+		err := fmt.Errorf("unsupported adopt mode %q", mode)
+		SetKonnectEntityAdoptedConditionFalse(ent, kcfgkonnect.KonnectEntitiesFailedToAdoptReason, err)
+		SetKonnectEntityProgrammedConditionFalse(ent, kcfgkonnect.KonnectEntitiesFailedToAdoptReason, err)
+		return err
+	}
+
+	var err error
+	switch obj := any(ent).(type) {
+	case *konnectv1alpha1.KonnectCloudGatewayNetwork:
+		err = adoptKonnectCloudGatewayNetworkMatch(ctx, sdk.GetCloudGatewaysSDK(), obj, adoptOpts.Konnect.ID)
```

Comment:
I think it is better to name them as `adoptKonnectCloudGatewayNetwork` and move the check of adopt mode there.